### PR TITLE
[MAINTENANCE] Prevent deserializing all data sources when adding a new one

### PR DESCRIPTION
**Pull request by cdkini**
To check for existence, we look into `context.data_sources.all()` - we only really need to check the key/name here. Let's use `self._datasources` `DatasourceDict` to do this.

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -535,7 +535,7 @@ def _add_fluent_datasource(
             )
 
         # We currently don't allow one to overwrite a datasource with this internal method
-        if datasource_name in self.data_sources.all():
+        if datasource_name in self._datasources:
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:538]: Nit - we should better differentiate `self._data_sources` and `self._datasources` - we should combine the two or perhaps make the latter `self._data_source_dict`

**Comment by TrangPham** [comment]
Thanks for your quick action here. There are other places which would have a similar issue. For example:
`AbstractDataContext.get_datasource` also calls `.data_sources.all()[name]`

As much as possible when an action pertains to one entity, the code shouldn't be verifying all other entities.

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 1 above)
Comment (TrangPham) [reply great_expectations/data_context/data_context/abstract_data_context.py:538]: May I suggest `_raw_data_sources`

**Review by rreinoldsc — APPROVED** [review]
Also seeing this bug, thanks for fixing

**Pull request closed without merging**